### PR TITLE
Refactor Terraform executor

### PR DESCRIPTION
**Pull request by johananl**
This PR paves the way towards https://github.com/kinvolk/lokomotive/issues/716. By making `terraform.Executor` accept an arbitrary series of "steps" we can now declaratively define the infrastructure pieces a platform is composed of.

Example:

```go
steps := []terraform.ExecutionStep{
    {
        Description: "create controllers",
        Args: []string{
            "apply",
            "-auto-approve",
            fmt.Sprintf("-target=module.packet-%s.packet_device.controllers", c.ClusterName),
        },
    },
    {
        Description: "create DNS records",
        Args:        []string{"apply", "-auto-approve", "-target=module.dns"},
    },
    {
        Description: "refresh Terraform state",
        Args:        []string{"refresh"},
    },
    {
        Description:      "complete infrastructure creation",
        Args:             []string{"apply", "-auto-approve"},
        // Verify DNS entries were manually created by the user before proceeding.
        PreExecutionHook: dns.ManualConfigPrompt(&c.DNS),
    },
}
```

I've deliberately avoided refactoring all the various platforms to keep the scope of this PR narrow. I've only refactored `platform/packet` (which is our most complex platform implementation at the moment) to demonstrate how the new structure can be used. The rest of the platforms will be refactored in subsequent PRs.

### CODE REVIEW

File: pkg/dns/dns.go [hunk 1]
```diff
@@ -61,41 +61,47 @@ func (c *Config) Validate() error {
...
-		var input string
-		fmt.Scanln(&input)
+			var input string
 
-		if input == "skip" {
-			break
-		} else if input != "" {
-			continue
-		}
+			_, err := fmt.Scanln(&input)
```
Comment (johananl) [inline pkg/dns/dns.go:81]: This breaks manual DNS. Working on a fix.

Comment (johananl) [reply pkg/dns/dns.go]: Done.

File: pkg/terraform/executor.go [hunk 2]
```diff
@@ -176,14 +207,27 @@ func tailFile(path string, done chan struct{}, wg *sync.WaitGroup) {
...
+// Execute accepts one or more ExecutionSteps and executes them sequentially in the order they were
+// provided. If a step has a PreExecutionHook defined, the hook is run prior to executing the step.
+// If any error is encountered, the error is returned and the execution is halted.
+func (ex *Executor) Execute(steps ...ExecutionStep) error {
+	for _, s := range steps {
+		if s.PreExecutionHook != nil {
+			ex.logger.Printf("Running pre-execution hook for step %q", s.Description)
+
+			if err := s.PreExecutionHook(ex); err != nil {
+				return fmt.Errorf("pre-execution hook failed: %w", err)
```
Comment (invidian) [inline pkg/terraform/executor.go:219]:
```suggestion
				return fmt.Errorf("running pre-execution hook: %w", err)
```

File: pkg/dns/dns.go [hunk 3]
```diff
@@ -61,37 +64,46 @@ func (c *Config) Validate() error {
...
-		var input string
-		fmt.Scanln(&input)
+			var input string
 
-		if input == "skip" {
-			break
-		} else if input != "" {
-			continue
-		}
+			reader := bufio.NewReader(os.Stdin)
```
Comment (invidian) [inline pkg/dns/dns.go:84]: It seems switch to NewReader should be done in separate commit?

Comment (johananl) [reply pkg/dns/dns.go:84]: I did this to satisfy the linter. We didn't actually touch this code in this PR, we just moved it around a bit but it was enough for the linter to complain. The complaint was valid though - we had an unhandled error from `fmt.Scanln()`.

Does it still make sense to you to split this?

Comment (johananl) [reply pkg/dns/dns.go:84]: Moved this to a separate commit to save time. I can re-squash if necessary.

Comment (invidian) [reply pkg/dns/dns.go:84]: Ah, I didn't know the change was driven by the linter. I think it make sense to do that in separate commit then, thanks!

File: pkg/dns/dns.go [hunk 4]
```diff
@@ -15,11 +15,14 @@
 package dns
```
Comment (invidian) [inline pkg/dns/dns.go:15]: By reading the commit message, I'm under the impression, it would be good to test such use cases (user blindly hitting Enter key etc.). But it's out of scope I guess.

File: pkg/dns/dns.go [hunk 5]
```diff
@@ -61,41 +64,50 @@ func (c *Config) Validate() error {
...
-// AskToConfigure reads the required DNS entries from a Terraform output,
-// asks the user to configure them and checks if the configuration is correct.
-func (c *Config) AskToConfigure(ex *terraform.Executor) error {
-	dnsEntries, err := readDNSEntries(ex)
-	if err != nil {
-		return err
-	}
+// ManualConfigPrompt returns a terraform.ExecutionHook which prompts the user to configure DNS
+// entries manually and verifies the entries were created successfully.
+func ManualConfigPrompt(c *Config) terraform.ExecutionHook {
```
Comment (invidian) [inline pkg/dns/dns.go:69]: Why change the receiver to the argument?

File: pkg/dns/dns.go [hunk 6]
```diff
@@ -61,41 +64,50 @@ func (c *Config) Validate() error {
...
+	return func(ex *terraform.Executor) error {
+		dnsEntries, err := readDNSEntries(ex)
+		if err != nil {
+			return err
+		}
 
-	fmt.Printf("Please configure the following DNS entries at the DNS provider which hosts %q:\n", c.Zone)
-	prettyPrintDNSEntries(dnsEntries)
+		fmt.Printf("Please configure the following DNS entries at the DNS provider which hosts %q:\n", c.Zone)
```
Comment (invidian) [inline pkg/dns/dns.go:76]: Perhaps we could have another unexported function to ask questions, which would take dnsEntries as an argument? That would simplify the testing.

File: pkg/dns/dns.go (hunk 4 above)
Comment (johananl) [reply pkg/dns/dns.go:15]: IMO - yes, and yes.

File: pkg/dns/dns.go (hunk 6 above)
Comment (johananl) [reply pkg/dns/dns.go:76]: What do you mean by a function for asking questions? Wrap `fmt.Printf()` in a function?
Anyway, IMO this belongs in a "refactor `pkg/dns`" task, i.e. out of scope.

File: pkg/dns/dns.go (hunk 5 above)
Comment (johananl) [reply pkg/dns/dns.go]: No good reason. Fixed.

File: pkg/dns/dns.go (hunk 6 above)
Comment (invidian) [reply pkg/dns/dns.go:76]: I meant, instead of taking executor as an argument and then execute the prints etc. it would be better to have a function like this, which takes an executor, gets dnsEntries using helper and passes that to function which does prints and reads user input. This way, we could fully test prints and reading user input.

Comment (johananl) [reply pkg/dns/dns.go:76]: OK, feel free to open an issue if you think we should do that. I'm resolving this since I think we agree this is out of scope here.

Comment (invidian) [reply pkg/dns/dns.go:76]: Not sure if such small things are worth creating an issue and allocating dedicated time. Seems more like a thing to do together with something IMO.
